### PR TITLE
Reset node playground workers when topics change.

### PR DESCRIPTION
**Pull request by defunctzombie**
**User-Facing Changes**
Node playground nodes are updated with generated types as the datasource loads new topics and datatypes.

**Description**
Node playground nodes generate code for messages and inputs from the set of available topics and datatypes. This code generation would run once on startup. If the generation happened to run before the data source provided the topics and datatypes, the user would see compilation errors in their nodes and the nodes would not run.

This change re-generates the nodes when the data source topics change. This allows the node code to compile and the nodes to start running as the data source loads available topics and datatypes.

Fixes: #2916

### CODE REVIEW

**Comment by defunctzombie** [comment]
I've removed the caching of node registrations as that was another layer that prevented re-building nodes when they should be rebuilt. It is true that nodes will be re-built when seeking or when topics change, but that seems like it happens quickly enough that its not a massive impact. We can introduce better "reset" behavior and caching if we do re-visit this code but it felt better to have things working as a tradeoff.

File: packages/studio-base/src/players/UserNodePlayer/index.ts [hunk 1]
```diff
@@ -644,6 +623,17 @@ export default class UserNodePlayer implements Player {
         await this._resetWorkers();
         this.setSubscriptions(this._subscriptions);
         this.requestBackfill();
+      } else {
+        // Reset node state after seeking
+        let shouldReset = activeData.lastSeekTime !== this._lastPlayerStateActiveData.lastSeekTime;
+
+        // reset if topics change
+        shouldReset ||= activeData.topics !== this._lastPlayerStateActiveData.topics;
```
Comment (jtbandes) [inline packages/studio-base/src/players/UserNodePlayer/index.ts:631]: Should we also check `datatypes`?

**Review by jtbandes — APPROVED** [review]
> It is true that nodes will be re-built when seeking or when topics change, but that seems like it happens quickly enough that its not a massive impact.

What about when scrubbing quickly (dragging the slider)? Seems like rebuilding nodes on every frame is probably a huge perf impact.

**Comment by defunctzombie** [comment]
I've thought about this and agree that re-compiling is probably not ideal. I'll bring back the caching and see if clearing it when topics changes is sufficient. If not, I might change the caching altogether since right now it looks like it stores old versions of code as well as new versions which seems unnecessary (rebuild on undo doesn't seem terrible).

**Comment by defunctzombie** [comment]
@jtbandes I reworked the logic to clear the cache when topics and datatypes change. I've left the cache in place otherwise.

File: packages/studio-base/src/players/UserNodePlayer/index.ts [hunk 2]
```diff
@@ -644,6 +640,23 @@ export default class UserNodePlayer implements Player {
...
+        // Reset node state after seeking
+        let shouldReset = activeData.lastSeekTime !== this._lastPlayerStateActiveData.lastSeekTime;
+
+        // When topics or datatypes change we also need to re-build the nodes so we clear the cache
+        if (
+          activeData.topics !== this._lastPlayerStateActiveData.topics ||
+          activeData.datatypes !== this._lastPlayerStateActiveData.datatypes
+        ) {
+          shouldReset ||= true;
+          this._nodeRegistrationCache = [];
```
Comment (jtbandes) [inline packages/studio-base/src/players/UserNodePlayer/index.ts:653]: It was initially confusing why this isn't under `shouldReset` but I see now that there are 2 levels of resetting: just resetting the node execution, vs re-compiling the nodes.

**Review by jtbandes — APPROVED** [review]

**Pull request merged**